server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -2157,26 +2157,29 @@ public boolean assertRetentionLeasesPersisted() throws IOException {
     /**
      * Syncs the current retention leases to all replicas.
      */
-    public void syncRetentionLeases() {
+    public void syncRetentionLeases(ActionListener<ReplicationResponse> listener) {
         assert assertPrimaryMode();
         verifyNotClosed();
         replicationTracker.renewPeerRecoveryRetentionLeases();
         final Tuple<Boolean, RetentionLeases> retentionLeases = getRetentionLeases(true);
-        if (retentionLeases.v1()) {
+        if (retentionLeases.v1() || listener != null) {
             logger.trace("syncing retention leases [{}] after expiration check", retentionLeases.v2());
-            retentionLeaseSyncer.sync(
-                shardId,
-                shardRouting.allocationId().getId(),
-                getPendingPrimaryTerm(),
-                retentionLeases.v2(),
-                ActionListener.wrap(
-                    r -> {},
+            if (listener == null) {
```

Comment:
Looks like using a NULL listener is used only once in productive code. Such I'd really prefer to always expect a listener here by moving this fallback to the place where a NULL would be used (`IndexServer.syncRetentionLeases()`).